olp-cpp-sdk-core/src/client/PendingUrlRequests.h

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) 2020 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#pragma once
+
+#include <map>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+
+#include "olp/core/client/CancellationContext.h"
+#include "olp/core/client/Condition.h"
+#include "olp/core/client/ErrorCode.h"
+#include "olp/core/client/HttpResponse.h"
+#include "olp/core/client/OlpClient.h"
+#include "olp/core/client/OlpClientSettings.h"
+#include "olp/core/http/HttpStatusCode.h"
+#include "olp/core/http/NetworkConstants.h"
+
+namespace olp {
+namespace client {
+
+/// This class represents one URL request holding one or more callbacks.
+class PendingUrlRequest {
+ public:
+  /// Alias for the CancellationContext execute function.
+  using ExecuteFuncType = std::function<CancellationToken(http::RequestId& id)>;
+  /// Alias for the CancellationContext cancel function.
+  using CancelFuncType = client::CancellationContext::CancelFuncType;
+
+  /// Identifies a invalid request Id.
+  static constexpr http::RequestId kInvalidRequestId =
+      static_cast<http::RequestId>(http::RequestIdConstants::RequestIdInvalid);
+  /// Cancelled Network request error code.
+  static constexpr int kCancelledStatus =
+      static_cast<int>(http::ErrorCode::CANCELLED_ERROR);
+
+  /// Append one callback to the request.
+  size_t Append(NetworkAsyncCallback callback);
+
+  /// Calls the managed CancellationContext ExecuteOrCancelled so that it can
+  /// remember the func returned CancellationToken to be able to cancel any
+  /// async operation taking place in our name. Should be called with the
+  /// Network trigger lambda.
+  bool ExecuteOrCancelled(const ExecuteFuncType& func,
+                          const CancelFuncType& cancel_func = nullptr);
+
+  /// Get the Network request Id associated with this request.
+  http::RequestId GetRequestId() const {
+    std::lock_guard<std::mutex> lock(mutex_);
+    return http_request_id_;
+  }
+
+  /// Cancels the ongoing Network request. Do not block for long!
+  /// NOTE: This should only be called if you know what you are doing.
+  void Cancel() { context_.CancelOperation(); }
```

Comment:
I would rather rename this to something different so there will be less confusion for the user between this and Cancel(callback_id)

Reply:
Will do.